app/services/bulk_unsubscribe_service.rb

```diff
@@ -0,0 +1,133 @@
+# rubocop:disable Metrics/BlockLength
+
+module BulkUnsubscribeService
+  def self.call(
+        content_ids_and_replacements,
+        subscriber_limit: 1_000_000,
+        courtesy_emails_every_nth_email: 500
+      )
+    affected_subscriber_list_ids = content_ids_and_replacements
+                                     .keys
+                                     .flat_map do |content_id|
+      SubscriberList.find_by_links_value(content_id).pluck(:id)
+    end
+
+    subscriptions_to_end = Subscription
+                             .where(subscriber_list_id: affected_subscriber_list_ids)
+                             .includes(:subscriber)
+                             .active
+
+    subscriptions_to_end
+      .group_by(&:subscriber)
+      .sort_by { |_subscriber, subscriptions| subscriptions.length * -1 } # start with most subscriptions
+      .take(subscriber_limit)
+      .each_with_index do |(subscriber, subscriptions), index|
+
+      subscription_details = subscriptions.map do |subscription|
+        relevant_content_id = subscription
+                                .subscriber_list
+                                .links
+                                .values
+                                .flatten
+                                .find do |links_content_id|
+          content_ids_and_replacements.key? links_content_id
+        end
+
+        [
+          subscription.subscriber_list.title,
+          content_ids_and_replacements.fetch(relevant_content_id)
+        ]
+      end
+
+      subscription_details = subscription_details.sort_by do |(title, _replacement)|
+        title
+      end
+
+      email = nil
+      ActiveRecord::Base.transaction do
+        email = send_email_to_subscriber(
```

Comment:
Should this be in this transaction? If the Transaction fails, the BulkUnsubscribeEmailBuilder will have already been called and the email queued.